app/provider_details/provider_service.py

```diff
@@ -36,42 +42,77 @@ def init_app(
 
     @property
     def strategies(self):
+        """ This is a dictionary with notification types as the keys. """
         return self._strategies
 
     def validate_strategies(self) -> None:
         for notification_type, strategy in self.strategies.items():
             strategy.validate(notification_type)
 
     def get_provider(self, notification: Notification) -> ProviderDetails:
-        template_or_service_provider_id = self._get_template_or_service_provider_id(notification)
-        if template_or_service_provider_id:
-            provider = get_provider_details_by_id(template_or_service_provider_id)
+        """
+        Return an instance of ProviderDetails that is appropriate for the given notification.
+        """
 
-            if provider is None:
-                raise InvalidProviderException(f'provider {template_or_service_provider_id} could not be found')
-            elif not provider.active:
-                raise InvalidProviderException(f'provider {template_or_service_provider_id} is not active')
+        # This is a UUID (ProviderDetails primary key).
+        provider_id = self._get_template_or_service_provider_id(notification)
+        logger.debug("notification = %s", notification)
+        logger.debug("provider_id = %s", provider_id)
```

Comment:
Does this actually work? It does not work for print statements. Old school print statements were:
`"notification = %s" % notification`

Reply:
Would this raise an exception if provider_id is None?

Reply:
It works.  The second parameter to logging functions is *args.  I looked at the CloudWatch logs after making these changes and saw the expected output.

There's no exception if provider_id is None because `str(None) --> 'None'`.  I also saw this in CloudWatch, and it lead me to the source of the user flows test failure.

Reply:
> it lead me to the source of the user flows test failure.
I'm curious to know what that source of the test failure was (a mini post-mortem if you will).